Add Break Handler

EnvisionCoding opened the pull request:
Breakhandler v1.0 release.

EnvisionCoding committed:
Break-handler Initial Commit
- Version 0.1
- Add basic file boilerplate
- Add Panel Icon
- Add Clock and Timer facets
- Add Developer name to PluginDescriptor
- Add Break Method types

HyperAlch committed:
Break Handler Method UI Refactor

HyperAlch committed:
Total restructure & Break Duration added

HyperAlch committed:
Break handler timer done

HyperAlch committed:
Runtime Duration & Run Timer done

HyperAlch committed:
Break Handler Account Creds done

HyperAlch committed:
Break Handler Two Factor Auth done

EnvisionCoding committed:
Update Break-handler Panels
- Version 0.2
- Update Panel Image
- Refactor JTextFields to FlatTextFields
- Add Panel Utilities
- Format timers and timer inputs

EnvisionCoding committed:
Break-handler generate random time durations
- Version 0.3
- Add dynamic break duration generation
- Add dynamic run time duration generation
- Add 2 Timer Types

EnvisionCoding committed:
Break-handler Automate Initial Durations
- Version 0.4
- Add Config for default break and run time durations.
- Generate initial break and run time timer values on startup
- Add accessors for panels

HyperAlch committed:
getMinimumTimeAmount & getMaximumTimeAmount access refactored

EnvisionCoding committed:
Break-handler Timer Logic
- Version 0.5
- Add working run time timer
- Add working break timer
- Add basic logic check to see which timer should be running
- Refactored TimerTypes enum's directory

HyperAlch committed:
Breakhandler: Account & Settings tabs completed

HyperAlch committed:
Breakhandler: Current Times section done

HyperAlch committed:
Breakhandler: Regenerate Times done

EnvisionCoding committed:
Break-handler State Management Phase 1
- Version 0.6
- Add Break Handler States
- Manage Break Handler States [Run and AFK_BREAK]
- Add config options for discord webhooks
- Grab break type from panel
- Temporarily reduced default timer durations for testing

HyperAlch committed:
Breakhandler: Titles Refactored

HyperAlch committed:
Breakhandler: setStyle() impl for components

HyperAlch committed:
Merge remote-tracking branch 'origin/breakhandler-ui' into breakhandler

# Conflicts:
#	runelite-client/src/main/java/net/runelite/client/plugins/envisionplugins/breakhandler/BreakHandlerPanel.java
#	runelite-client/src/main/java/net/runelite/client/plugins/envisionplugins/breakhandler/ui/breakmethod/BreakMethodPanel.java
#	runelite-client/src/main/java/net/runelite/client/plugins/envisionplugins/breakhandler/ui/breakmethod/BreakMethodParentPanel.java

HyperAlch committed:
Breakhandler: Comments added to point out tabbed panel groupings

HyperAlch committed:
Breakhandler: Header adjusted + New Discord Icon

EnvisionCoding committed:
Break-handler Debugging QoL
- Version 0.7
- Introduce debug toggle via config
- Limit debug messages

EnvisionCoding committed:
Break-handler Convert to New Timers
- Version 0.8
- Refactor logic to update new tabbed timers
- Remove old timers

EnvisionCoding committed:
Breakhandler: Impl Regenerate Timer buttons
- Version 0.9
- Add button logic to regenerate run time duration
- Add button logic to regenerate break time duration

EnvisionCoding committed:
Breakhandler: Impl Break Now Btn
- Version 0.10
- Add button logic to break now button

HyperAlch committed:
Breakhandler: regenerateRunTimeBtn fix

HyperAlch committed:
Breakhandler: Regen buttons fixed

EnvisionCoding committed:
Merge branch 'chsami:main' into breakhandler

EnvisionCoding committed:
Breakhandler: Fixed bug dealing with regenerating run timer
- Version 0.11
- Prevent user from regenerating run timer during a break
- Add idle message during break
- Wrap main loop in try catch to be more inline with best practice

EnvisionCoding committed:
Breakhandler: Fixed bug with unfocus on min and max timers
- Version 0.12
- Removed calculating expected timers when changing the min and max durations
- Add calc timers to startup

HyperAlch committed:
Discord Webhooks Finished

HyperAlch committed:
Left over comments removed

EnvisionCoding committed:
Breakhandler: Impl discord webhooks
- Version 0.13
- Add plugin name to discord messages
- Add fields for parent plugin to control webhook messages
- Add config field for client name

EnvisionCoding committed:
Breakhandler: Bug fix rename webhook field title
- Version 0.14
- Fix field title for Plugin name

HyperAlch committed:
Merge remote-tracking branch 'origin/breakhandler' into development

# Conflicts:
#	runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/walker/Walker.java

HyperAlch committed:
Breakhandler: get methods for Account tab added

HyperAlch committed:
Breakhandler: getPasswordEncryptedValue method added

HyperAlch committed:
Sync with main repo

HyperAlch committed:
Merge branch 'chsami-development' into development

HyperAlch committed:
Breakhandler: World hopping IU added

EnvisionCoding committed:
Breakhandler: Impl logout break Phase 1
- Version 0.15
- Add ability to logout to break
- Limit number of discord notification when logging back in
- Impl logging back in post break (phase 1)
- Added ability to world hop after each break
- Added ability to pick world types to hop to post break
- Added ability to pick world region types post break world hop
- Update switchToLogout Microbot method to work with multiple screen display settings

HyperAlch committed:
Breakhandler: Disable F2A & Pin (For now)

HyperAlch committed:
Merge branch 'development' of https://github.com/EnvisionCoding/microbot into development

EnvisionCoding committed:
Breakhandler: Impl and fix breaks Phase 2
- Version 0.16
- Fix AFK break not actually preventing script from running
- Add Microbot status for AFK break
- Fix run time counting down on login screen
- Cleanup some debug statements
- Setup Nate's pie shell maker to use break handler
- Setup Nate's power miner to use break handler

HyperAlch committed:
Breakhandler: TimeManager introduction

HyperAlch committed:
Breakhandler: breakTime old system temp fix

HyperAlch committed:
Breakhandler: Major refactor & Time Manager

HyperAlch committed:
Breakhandler: Use Optional to keep the breakTimer from activating while waiting

HyperAlch committed:
Breakhandler: Fixed regen visual glitch

EnvisionCoding committed:
Breakhandler: Impl Notification Manager
- Version 0.17
- Add Notification Manager
- Convert old logging and discord notifications to use the new Manager
- Clean up commented out code

HyperAlch committed:
Breakhandler: AFK break fixed

HyperAlch committed:
Breakhandler: State cleanup

HyperAlch committed:
Breakhandler: Deleting old unused methods and fields

HyperAlch committed:
Breakhandler: Delete old Timer and Clock classes

HyperAlch committed:
Breakhandler: State refactor

HyperAlch committed:
Breakhandler: Reorginize BreakHandlerScript

EnvisionCoding committed:
Breakhandler: Impl Failure State Phase 1
- Version 0.23
- Refactor NotificationManager limiting logic and readability
- Introduce the failure state
- If no parent plugin is running change state to failure
- If no account creds are given change state to failure

EnvisionCoding committed:
Breakhandler: Impl Failure State Phase 1
- Version 0.24
- Fix Rs2Tab not working with every display setting

HyperAlch committed:
Reset back to 77f6c7a

HyperAlch committed:
Sync with main repo

HyperAlch committed:
Merge branch 'development' of https://github.com/EnvisionCoding/microbot into development

EnvisionCoding committed:
Breakhandler: Update Rs2Tab switchToLogout for all display settings
- Version 0.23
- Fix Rs2Tab not working with every display setting

EnvisionCoding committed:
Breakhandler: Introduce FAILURE state
- Version 0.24
- Enter failure state if no compatible plugins are running
- Remove old compatible check

HyperAlch committed:
Breakhandler: Failure UI is ready for implementation

HyperAlch committed:
Breakhandler: Error display system done

HyperAlch committed:
Sync with official repo

EnvisionCoding committed:
Breakhandler: Improve Parent Plugin compatibility
- Version 0.26
- Increase default break timers via plugin configuration
- Add single line init function for the break handler for lazy developers
- Clean up deprecated code

EnvisionCoding committed:
Breakhandler: Missing account creds failure state
- Version 0.27
- Check for missing account creds and throw failure state

HyperAlch committed:
Breakhandler: Error repaint fix & Invalid Username / Password error added

HyperAlch committed:
Breakhandler: Password valid check fixed

EnvisionCoding committed:
Breakhandler: Clean up code for release
- Version 0.28
- Remove discord notification if missing account creds
- Fix null error in Rs2Tab

HyperAlch committed:
Merge branch 'development' of https://github.com/chsami/microbot into development

EnvisionCoding committed:
Breakhandler: Release
- Version 1.0

chsami commented on runelite-client/src/main/java/net/runelite/client/plugins/nateplugins/skilling/nateminer/nateminer/MiningScript.java:
```diff
@@ -18,15 +19,34 @@ public boolean run(MiningConfig config) {
+                 *      Then notify that a break can be started
+                 *      Finally lets wait until the break is over.
+                 */
+                if (BreakHandlerScript.getHasRunTimeTimerFinished()) {
```
Can this logic live inside the breakHandlerScript? I just want to avoid editing 50 scripts because something inside this code block has changed

chsami commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/tabs/Rs2Tab.java:
```diff
@@ -160,10 +161,34 @@ public static boolean switchToMusicTab() {
+        int logout_widget_id;
+
+        try {
+            if (Rs2Widget.getWidget(FIXED_CLASSIC_DISPLAY__FIXED_VIEWPORT_OPTIONS_TAB) != null) {
```
try to split out the logic for getting the logout widget and actually logging out.

```
public static boolean switchToLogout() {
    if (getCurrentTab() == InterfaceTab.LOGOUT) return true;

    int logout_widget_id = getLogoutWidgetId();
    
    if (logout_widget_id == 0) return false;

    Widget tab = Microbot.getClient().getWidget(logout_widget_id);
    if (tab == null) return false;
    
    Microbot.getMouse().click(tab.getBounds());
    sleep(200, 600);
    
    return getCurrentTab() == InterfaceTab.LOGOUT;
}

private static int getLogoutWidgetId() {
    try {
        if (Rs2Widget.getWidget(FIXED_CLASSIC_DISPLAY__FIXED_VIEWPORT_OPTIONS_TAB) != null) {
            return FIXED_CLASSIC_DISPLAY__FIXED_VIEWPORT_OPTIONS_TAB;
        } else if (Rs2Widget.getWidget(RESIZABLE_CLASSIC_DISPLAY__RESIZABLE_VIEWPORT_LOGOUT_ICON) != null) {
            return RESIZABLE_CLASSIC_DISPLAY__RESIZABLE_VIEWPORT_LOGOUT_ICON;
        } else if (Rs2Widget.getWidget(RESIZABLE_MODERN_DISPLAY__RESIZABLE_VIEWPORT_BOTTOM_LINE_OPTIONS_ICON) != null) {
            return RESIZABLE_MODERN_DISPLAY__RESIZABLE_VIEWPORT_BOTTOM_LINE_OPTIONS_ICON;
        }
    } catch (Exception ignored) {
        // Rs2Widget.getWidget returns null if the game isn't finished loading
        // Consider logging the exception for debugging purposes
    }

    return 0;
}
```

EnvisionCoding replied:
Could you elaborate, this should be just switching to the required tab. I only extended the original method to account for all three display settings.

chsami replied:
SwitchToLogout and getting the logout widget are two seperate tasks, and it is better to split them up. code easier to read and maintain and won't violate the SRP principle.

chsami commented on runelite-client/src/main/java/net/runelite/client/plugins/envisionplugins/breakhandler/BreakHandlerPlugin.java:
```diff
@@ -0,0 +1,101 @@
+        name = PluginDescriptor.Envision + "Break Handler",
+        description = "Break Handler for Microbot",
+        tags = {"microbot", "utility"},
+        enabledByDefault = true
```
not sure if this is needed here, but do we want the BreakHandlerPlugin to always be on? 

if yes, than you can use the         alwaysOn = true  property

HyperAlch committed:
Breakhandler: BreakHandlerExecutor impl

EnvisionCoding committed:
Breakhandler: PR fix for Rs2Tab
- Version 1.1
- Followed SRP role by splitting out the switchToMethod

EnvisionCoding committed:
Breakhandler: Fix login and no longer enabled by default
- Version 1.2
- No longer defaults to EnabledByDefault
- Add logic checks to see if break handler is enabled
- Created a separate Login class for break handler usage
